### ISSUE TITLE
Non-Parametric Outlier Synthesis (ICLR 2023)

### ISSUE THREAD

**Issue by djamahl99** [Luoyadan/Research-Group-Progress#12]
![image](https://user-images.githubusercontent.com/19633934/228407455-6aa8c1d2-c2a7-412a-885d-68906a339b8f.png)

# Introduction and Background

- OOD detection is important for reducing FPs on OOD data, important for reducing risks in safety critical problems, such as autonomous driving
- Models often produce overconfident predictions on unseen data
- Outlier synthesis can be used to fake real latent vectors, but previously have relied on strong assumptions, such as Gaussian distribution (e.g. VOS [same authors] does OOD detection and reweighting in classification, seg, and detection using Virtual Outlier Synthesis with the gaussian assumption
- Rejection sampling can allow us to sample from difficult to analyse distributions (they follow textbook from Kroese, a UQ statistics lecturer 😄  )

# Method
- Define the lower level set for  OOD samples: $\mathcal{L} := \{ x: \mathcal{P}_{in} \leq \beta \}$.
- Define $\mathcal{P}_{in}(x)$ as the probability that a data sample, $x$, is in distribution. 
- $\beta$ is a set parameter for our desired proportion of OOD samples, set so that the false OOD detection rate is 0.05, to elaborate $0.05 = \int_{\mathcal{L}} \mathcal{P_{in}} (x) dx $. 
- We classify a sample as OOD if $x \in \mathcal{L}$
- OOD conditional distribution $Q(x | OOD) = \frac{\mathcal{1} [\mathcal{P}_{in}(x) \leq \beta] \mathcal{P}_{in}(x)} {1 - \mathcal{z}_{out}}$

## Steps:
1. Use nearest neighbors to estimate the probability density of points in the training sample
2.  Estimate which points are outliers using $\beta$
3. Add Gaussian noise centered at randomly sampled outliers to create synthetic outliers 
4. Apply rejection sampling so the synthetic outliers are in the desired distribution, following the acceptance-rejection method from our very own UQ stats lecturer (see Rejection Sampling)
5. Use nearest neighbours again for classification and outlier detection

### Submethod - Rejection Sampling (steps 3 & 4 elaborated):
1. Draw an index in $\mathcal{D}_{in}$ by $i \sim uniform[n]$ where $n$ is the number of training samples
2. Draw sample $v$ (candidate synthesised outlier) in the feature space from a Gaussian kernel centered at $h(x_i)$ with covariance $\sigma^2 \mathbb{I}: v \sim \mathcal{N}(h(x_i), \sigma^2 \mathbb{I}) $
3. Accept $v$ with probability $\frac{Q(h(x) | OOD)}{M \mathcal{P}_{in}}$ where $M$ is an upper bound on the conditional likelihood ratio of OOD $Q(h(x) | OOD) / \mathcal{P}_{in}$

## Test-time OOD Detection
![image](https://user-images.githubusercontent.com/19633934/228410190-e2d149b1-214b-428c-a20a-93b8e295fbf8.png)

### Algorithm

![image](https://user-images.githubusercontent.com/19633934/228409109-c0ba3a83-b4fc-46ec-a0cc-1b1bba328ef0.png)

### Risk
![image](https://user-images.githubusercontent.com/19633934/228407982-89edd36d-4b61-45ce-a601-cb83adff2069.png)

# Experiments

### Datasets
- CIFAR100 and ImageNet as ID
- iNaturalist, SUN, Places, Texture as OOD

### Models
- pretrained CLIP ViT, and fine-tune the last two blocks of the image encoder 
- also do an ablation with the RN50 CLIP encoder

## Results
VOS is their previous method, Virtual Outlier Synthesis, which uses a Gaussian distribution on the latent space. It firsts models the feature embedding as a mixtue of multivariate Gaussians, and then samples the virtual outliers from the low-likelihood region in the embedding space. For VOS+, they use the loss function of NPOS, but leave the sampling as a mixture of parametric multivariate Gaussians. VOS+ and NPOS only differ in how outliers are synthesised.

![image](https://user-images.githubusercontent.com/19633934/228406415-f8a16fe6-84aa-4431-96c9-44d35d87a129.png)

We can see that by comparing VOS+ and NPOS, (parametric vs non-parametric sampling), that parametric sampling provides better performance all around. The updated training loss further shows a significant improvement for model regularization, with both parametric and non-parametric sampling. Furthermore, 

![image](https://user-images.githubusercontent.com/19633934/228407377-cfd4db8b-025c-4ab1-a9da-eab3e7f99e4e.png)
